Add LogfmtRenderer

airwoodix opened the pull request:
# Summary

This PR adds a `LogfmtRenderer` processor to format log lines in the [logfmt](https://brandur.org/logfmt) format.

The core of the work is done by the [`logfmt`](https://pypi.org/project/logfmt/) library, which is added as global dependency. Or should this be changed to feature-based optional?

Refactored the code in `KeyValueRenderer` to share the `sort_keys`/`key_order`/`drop_missing` logic. Tests are directly adapted from `KeyValueRenderer`.

Closes #322 


# Pull Request Check List

This is just a friendly reminder about the most common mistakes.  Please make sure that you tick all boxes.  But please read our [contribution guide](https://github.com/hynek/structlog/blob/main/.github/CONTRIBUTING.md) at least once, it will save you unnecessary review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing left to do.

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.
    - [x] New functions/classes have to be added to `docs/api.rst` by hand.
    - [x] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded).  Find the appropriate next version in our [`__init__.py`](https://github.com/hynek/structlog/blob/main/src/structlog/__init__.py) file.
- [x] Changes (and possible deprecations) are documented in the [changelog](https://github.com/hynek/structlog/blob/main/CHANGELOG.rst).

If you have *any* questions to *any* of the points above, just **submit and ask**!  This checklist is here to *help* you, not to deter you from contributing!

hynek commented on src/structlog/processors.py:
```diff
@@ -132,6 +99,86 @@ def __call__(
+    """
+    Return a function to sort items from an `event_dict`.
+
+    See :class:`KeyValueRenderer` for an explanation of the parameters.
```
```suggestion
    See `KeyValueRenderer` for an explanation of the parameters.
```

hynek commented on src/structlog/processors.py:
```diff
@@ -132,6 +99,86 @@ def __call__(
+    drop_missing: bool,
+) -> Callable[[EventDict], List[Tuple[str, Any]]]:
+    """
+    Return a function to sort items from an `event_dict`.
```
We have auto-linking for single-ticks.

```suggestion
    Return a function to sort items from an ``event_dict``.
```

hynek commented on docs/api.rst:
```diff
@@ -232,6 +232,16 @@ Please see :doc:`thread-local` for details.
+   .. doctest::
+
+      >>> from structlog.processors import LogfmtRenderer
+      >>> LogfmtRenderer(sort_keys=True)(None, None, {"a": 42, "b": [1, 2, 3]})
+      'a=42 b="[1, 2, 3]"'
+      >>> LogfmtRenderer(key_order=["b", "a"])(None, None,
+      ...                                      {"a": 42, "b": [1, 2, 3]})
+      'b="[1, 2, 3]" a=42'
```
Let's simplify

```suggestion
      >>> event_dict = {"a": 42, "b": [1, 2, 3]}
      >>> LogfmtRenderer(sort_keys=True)(None, None, event_dict)
      'a=42 b="[1, 2, 3]"'
      >>> LogfmtRenderer(key_order=["b", "a"])(None, None, event_dict)
      'b="[1, 2, 3]" a=42'
```

hynek commented on src/structlog/processors.py:
```diff
@@ -132,6 +99,86 @@ def __call__(
+        key_order: Optional[Sequence[str]] = None,
+        drop_missing: bool = False,
+    ):
+        self._ordered_items = _items_sorter(sort_keys, key_order, drop_missing)
```
if you like, you could add an `if logfmt is None: raise ...` here (if you're willing to add a test too).

airwoodix replied:
Fixed by 44a19e6 (drop `logfmt` dependency).

hynek commented on tests/test_processors.py:
```diff
@@ -125,6 +126,95 @@ def test_repr_native_str(self, rns):
         assert 2 == cnt
 
 
+class TestLogfmtRenderer:
```
Try to import logfmt above and then protect this whole class:

```suggestion
@pytest.mark.skipif(logfmt is None, reason="Needs logfmt package.")
class TestLogfmtRenderer:
```

airwoodix replied:
Fixed by 44a19e6 (drop `logfmt` dependency).

hynek reviewed (changes requested):
Wow, this is great thank you! logfmt doesn't seem to be too well-maintained these days, but we can give it a try before we implement it ourselves.

As you've suspected, we need to make logfmt optional, I've added instructions to the pyproject.toml lines

Tasks that I couldn't add to the diff:

- add test env with logfmt to tox.ini (analogous to the rich env)
- add a usage example to `typing_examples.py`

hynek commented:
Um, so I've had a look at logfmt (of course one second after sending the review 🤦‍♂️) and I don't think we need a dependency for [17 LoC](https://github.com/wlonk/logfmt-python/blob/9f7cd50355bc13ddefab4c48dbdc459d8700648c/logfmt/formatter.py#L5-L21)?

```python
def format_line(extra):
    outarr = []
    for k, v in extra.items():
        if v is None:
            outarr.append("%s=" % k)
            continue


        if isinstance(v, bool):
            v = "true" if v else "false"
        elif isinstance(v, numbers.Number):
            pass
        else:
            if isinstance(v, (dict, object)):
                v = str(v)
            v = '"%s"' % v.replace('"', '\\"')
        outarr.append("%s=%s" % (k, v))
    return " ".join(outarr)
```

airwoodix commented on src/structlog/processors.py:
```diff
@@ -25,6 +25,8 @@
     Union,
 )
 
+import logfmt
```
Fixed by 44a19e6 (drop `logfmt` dependency).

airwoodix commented on pyproject.toml:
```diff
@@ -10,7 +10,7 @@ name = "structlog"
 dynamic = ["version", "description"]
 requires-python = ">=3.6"
-dependencies = ["typing-extensions; python_version<'3.8'"]
+dependencies = ["typing-extensions; python_version<'3.8'", "logfmt>=0.4"]
```
Fixed by 44a19e6 (drop `logfmt` dependency).

airwoodix commented:
Thanks for the fast and detailed review!

> I don't think we need a dependency for 17 LoC?

Agreed. I added the logfmt rendering logic to `LogfmtRenderer.__call__` in 44a19e6.

Since there's no formal specification, I oriented myself to the behavior of the [go](https://pkg.go.dev/github.com/kr/logfmt) and [node](https://github.com/csquared/node-logfmt) implementations. The go docs have an example and a pseudo-spec.

I also quickly checked that Grafana's LogQL `logfmt` filter/parser is happy with quoted/unquoted strings, empty values (`key=`) and boolean flags (`key`). 

>    add test env with logfmt to tox.ini (analogous to the rich env)

I guess that this is obsolete with the dependency to `logfmt` removed?

>    add a usage example to typing_examples.py

I added a block similar to the ones for `JSONRenderer` in f3b5ffe, though I'm not sure how the `typing_examples.py` mechanism works.

airwoodix commented on src/structlog/processors.py:
```diff
@@ -132,6 +99,86 @@ def __call__(
+            return ""
+
+
+def _items_sorter(
```
Would you prefer integrating this into this PR or rather tackle it later?

hynek replied:
definitely later, I was mostly ordering my thoughts

hynek commented:
> Thanks for the fast and detailed review!

Thanks for the quick turnaround!

> > I don't think we need a dependency for 17 LoC?
> 
> Agreed. I added the logfmt rendering logic to `LogfmtRenderer.__call__` in [44a19e6](https://github.com/hynek/structlog/commit/44a19e6b999f9ee68e38189ba940722335d0424f).
> 
> Since there's no formal specification, I oriented myself to the behavior of the [go](https://pkg.go.dev/github.com/kr/logfmt) and [node](https://github.com/csquared/node-logfmt) implementations. The go docs have an example and a pseudo-spec.

Wonderful, looks great! We might want to add a `default` mechanism like json has, but good enough start!
 
> > add test env with logfmt to tox.ini (analogous to the rich env)
> I guess that this is obsolete with the dependency to `logfmt` removed?

yep

> > add a usage example to typing_examples.py
> I added a block similar to the ones for `JSONRenderer` in [f3b5ffe](https://github.com/hynek/structlog/commit/f3b5ffe1ea2d6ae99d914ab7e29b61c2c56ae196), though I'm not sure how the `typing_examples.py` mechanism works.

it just uses the APIs and looks whether mypy is happy or not. the code is never executed.
